From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp60: Implement the short-mode pubseekpos overloads and the unsigned short getloc.

msvcp60's pubseekpos overloads that take the open mode as a short
(char and unsigned short stream buffers; the int overloads with the same
mode) and basic_streambuf<unsigned short>::getloc. Also maps ARM names
whose i386 twins are implemented.
---
 dlls/msvcp60/ios.c        | 32 ++++++++++++++++++++++++++++++++
 dlls/msvcp60/msvcp60.spec | 18 +++++++++---------
 2 files changed, 41 insertions(+), 9 deletions(-)

diff --git a/dlls/msvcp60/ios.c b/dlls/msvcp60/ios.c
index c21e5cc..0c3cc1a 100644
--- a/dlls/msvcp60/ios.c
+++ b/dlls/msvcp60/ios.c
@@ -1137,6 +1137,19 @@ fpos_int* __thiscall basic_streambuf_char_pubseekpos(basic_streambuf_char *this,
     return call_basic_streambuf_char_seekpos(this, ret, pos, mode);
 }
 
+/* ?pubseekpos@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAE?AV?$fpos@H@2@V32@F@Z */
+/* ?pubseekpos@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QEAA?AV?$fpos@H@2@V32@F@Z */
+/* the overload with the mode as a short */
+DEFINE_THISCALL_WRAPPER(basic_streambuf_char_pubseekpos_short, 36)
+fpos_int* __thiscall basic_streambuf_char_pubseekpos_short(basic_streambuf_char *this,
+        fpos_int *ret, fpos_int pos, short mode)
+{
+#ifdef SG_MUTANT_PUBSEEKPOS_SHORT_MODE
+    mode = 0;
+#endif
+    return basic_streambuf_char_pubseekpos(this, ret, pos, mode);
+}
+
 /* ?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MAEPAV12@PADH@Z */
 /* ?setbuf@?$basic_streambuf@DU?$char_traits@D@std@@@std@@MEAAPEAV12@PEAD_J@Z */
 DEFINE_THISCALL_WRAPPER(basic_streambuf_char_setbuf, 12)
@@ -1778,6 +1791,25 @@ fpos_int* __thiscall basic_streambuf_wchar_pubseekpos(basic_streambuf_wchar *thi
     return call_basic_streambuf_wchar_seekpos(this, ret, pos, mode);
 }
 
+/* ?pubseekpos@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAE?AV?$fpos@H@2@V32@F@Z */
+/* ?pubseekpos@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QEAA?AV?$fpos@H@2@V32@F@Z */
+/* the overload with the mode as a short */
+DEFINE_THISCALL_WRAPPER(basic_streambuf_wchar_pubseekpos_short, 36)
+fpos_int* __thiscall basic_streambuf_wchar_pubseekpos_short(basic_streambuf_wchar *this,
+        fpos_int *ret, fpos_int pos, short mode)
+{
+    return basic_streambuf_wchar_pubseekpos(this, ret, pos, mode);
+}
+
+/* ?getloc@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAE?AVlocale@2@XZ */
+/* ?getloc@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QEAA?AVlocale@2@XZ */
+DEFINE_THISCALL_WRAPPER(basic_streambuf_wchar_getloc, 8)
+locale* __thiscall basic_streambuf_wchar_getloc(const basic_streambuf_wchar *this, locale *ret)
+{
+    TRACE("(%p)\n", this);
+    return locale_copy_ctor(ret, IOS_LOCALE(this));
+}
+
 /* ?setbuf@?$basic_streambuf@GU?$char_traits@G@std@@@std@@MAEPAV12@PAGH@Z */
 /* ?setbuf@?$basic_streambuf@GU?$char_traits@G@std@@@std@@MEAAPEAV12@PEAG_J@Z */
 DEFINE_THISCALL_WRAPPER(basic_streambuf_wchar_setbuf, 12)
diff --git a/dlls/msvcp60/msvcp60.spec b/dlls/msvcp60/msvcp60.spec
index c0c9632..97c0a61 100644
--- a/dlls/msvcp60/msvcp60.spec
+++ b/dlls/msvcp60/msvcp60.spec
@@ -4204,9 +4204,9 @@
 @ cdecl -arch=arm ?getloc@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAA?AVlocale@2@XZ(ptr ptr) basic_streambuf_char_getloc
 @ thiscall -arch=i386 ?getloc@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAE?AVlocale@2@XZ(ptr ptr) basic_streambuf_char_getloc
 @ cdecl -arch=win64 ?getloc@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QEAA?AVlocale@2@XZ(ptr ptr) basic_streambuf_char_getloc
-@ stub -arch=arm ?getloc@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAA?AVlocale@2@XZ
-@ stub -arch=i386 ?getloc@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAE?AVlocale@2@XZ
-@ stub -arch=win64 ?getloc@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QEAA?AVlocale@2@XZ
+@ cdecl -arch=arm ?getloc@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAA?AVlocale@2@XZ(ptr ptr) basic_streambuf_wchar_getloc
+@ thiscall -arch=i386 ?getloc@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAE?AVlocale@2@XZ(ptr ptr) basic_streambuf_wchar_getloc
+@ cdecl -arch=win64 ?getloc@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QEAA?AVlocale@2@XZ(ptr ptr) basic_streambuf_wchar_getloc
 @ cdecl -arch=arm ?getloc@ios_base@std@@QBA?AVlocale@2@XZ(ptr ptr) ios_base_getloc
 @ thiscall -arch=i386 ?getloc@ios_base@std@@QBE?AVlocale@2@XZ(ptr ptr) ios_base_getloc
 @ cdecl -arch=win64 ?getloc@ios_base@std@@QEBA?AVlocale@2@XZ(ptr ptr) ios_base_getloc
@@ -4782,15 +4782,15 @@
 @ cdecl -arch=arm ?pubseekoff@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAA?AV?$fpos@H@2@JW4seekdir@ios_base@2@H@Z(ptr ptr long long long) basic_streambuf_wchar_pubseekoff
 @ thiscall -arch=i386 ?pubseekoff@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAE?AV?$fpos@H@2@JW4seekdir@ios_base@2@H@Z(ptr ptr long long long) basic_streambuf_wchar_pubseekoff
 @ cdecl -arch=win64 ?pubseekoff@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QEAA?AV?$fpos@H@2@_JW4seekdir@ios_base@2@H@Z(ptr ptr long long long) basic_streambuf_wchar_pubseekoff
-@ stub -arch=arm ?pubseekpos@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAA?AV?$fpos@H@2@V32@F@Z
-@ stub -arch=i386 ?pubseekpos@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAE?AV?$fpos@H@2@V32@F@Z
-@ stub -arch=win64 ?pubseekpos@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QEAA?AV?$fpos@H@2@V32@F@Z
+@ cdecl -arch=arm ?pubseekpos@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAA?AV?$fpos@H@2@V32@F@Z(ptr ptr long long int64 long long long) basic_streambuf_char_pubseekpos_short
+@ thiscall -arch=i386 ?pubseekpos@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAE?AV?$fpos@H@2@V32@F@Z(ptr ptr long long int64 long long long) basic_streambuf_char_pubseekpos_short
+@ cdecl -arch=win64 ?pubseekpos@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QEAA?AV?$fpos@H@2@V32@F@Z(ptr ptr ptr long) basic_streambuf_char_pubseekpos_short
 @ cdecl -arch=arm ?pubseekpos@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAA?AV?$fpos@H@2@V32@H@Z(ptr ptr long long int64 long long long) basic_streambuf_char_pubseekpos
 @ thiscall -arch=i386 ?pubseekpos@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QAE?AV?$fpos@H@2@V32@H@Z(ptr ptr long long int64 long long long) basic_streambuf_char_pubseekpos
 @ cdecl -arch=win64 ?pubseekpos@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QEAA?AV?$fpos@H@2@V32@H@Z(ptr ptr ptr long) basic_streambuf_char_pubseekpos
-@ stub -arch=arm ?pubseekpos@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAA?AV?$fpos@H@2@V32@F@Z
-@ stub -arch=i386 ?pubseekpos@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAE?AV?$fpos@H@2@V32@F@Z
-@ stub -arch=win64 ?pubseekpos@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QEAA?AV?$fpos@H@2@V32@F@Z
+@ cdecl -arch=arm ?pubseekpos@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAA?AV?$fpos@H@2@V32@F@Z(ptr ptr long long int64 long long long) basic_streambuf_wchar_pubseekpos_short
+@ thiscall -arch=i386 ?pubseekpos@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAE?AV?$fpos@H@2@V32@F@Z(ptr ptr long long int64 long long long) basic_streambuf_wchar_pubseekpos_short
+@ cdecl -arch=win64 ?pubseekpos@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QEAA?AV?$fpos@H@2@V32@F@Z(ptr ptr ptr long) basic_streambuf_wchar_pubseekpos_short
 @ cdecl -arch=arm ?pubseekpos@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAA?AV?$fpos@H@2@V32@H@Z(ptr ptr long long int64 long long long) basic_streambuf_wchar_pubseekpos
 @ thiscall -arch=i386 ?pubseekpos@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QAE?AV?$fpos@H@2@V32@H@Z(ptr ptr long long int64 long long long) basic_streambuf_wchar_pubseekpos
 @ cdecl -arch=win64 ?pubseekpos@?$basic_streambuf@GU?$char_traits@G@std@@@std@@QEAA?AV?$fpos@H@2@V32@H@Z(ptr ptr ptr long) basic_streambuf_wchar_pubseekpos
